File: src/engine.ts
/**
 * The reading pipeline, driven from Swift instead of from a page.
 *
 * WHY THIS FILE EXISTS. The iOS app is SwiftUI, and the reading is not
 * reimplemented in Swift -- every figure in HANDOFF.md was measured against the
 * TypeScript in `src/lib/`, and a second implementation would be a second set
 * of numbers to keep in step with. So the phone runs this: the same modules the
 * desktop tool runs, in an off-screen WKWebView, with the interface stripped
 * off. `main.ts` is the browser's front end and this is the phone's back end;
 * neither owns an algorithm.
 *
 * THE PROTOCOL. Swift calls `window.tally.dispatch(json)` and gets everything
 * back through `webkit.messageHandlers.tally`, because the return value of
 * `evaluateJavaScript` cannot carry a promise. Every message is
 * `{ id, ok, result }` or `{ id, ok: false, error }`, except progress, which is
 * `{ event: "progress", ... }` and carries no id.
 *
 * WHAT DOES NOT CROSS THE BRIDGE. Page images. A letter page at 200 DPI is
 * 3.7MB of grayscale and a 116-page scan is 430MB of them; they are cut into
 * row crops here and dropped, exactly as `processFile` does in `main.ts`, and
 * Swift asks for one cell picture at a time by card and row. The PDF comes the
 * other way for the same reason -- Swift writes it to the inbox and sends a
 * URL, rather than a base64 string the size of the scan.
 */

import { loadCellMap, type CellMap } from "./lib/cells";
import { decodeModel, type DigitModel } from "./lib/digits";
import {
  assembleCard,
  cellsForSide,
  type ExtractedCard,
  type ExtractedCell,
  type PageCells,
  viewRect,
} from "./lib/extract";
import { cropToCanvas, toGray, type GrayImage } from "./lib/image";
import { isAutoAccepted, prefillFor } from "./lib/prefill";
import { rasterizePdf } from "./lib/pdf";
import {
  pairIntoCards,
  referenceTargets,
  registerAgainstBestSide,
  MIN_BANNER_OVERLAP,
  type PairingProblem,
} from "./lib/register";
import { fillTemplate, suggestFilename } from "./lib/xlsx";
import type { ExtractedCard as ExportCard } from "./lib/xlsx";

// ---------------------------------------------------------------------------
// The bridge
// ---------------------------------------------------------------------------

interface Bridge {
  postMessage(body: string): void;
}

declare global {
  interface Window {
    webkit?: { messageHandlers?: { tally?: Bridge } };
    tally: { dispatch(json: string): void };
  }
}

function post(message: unknown): void {
  const bridge = window.webkit?.messageHandlers?.tally;
  const body = JSON.stringify(message);
  // Without a host the engine still runs -- `npm run dev` and engine.html in a
  // desktop browser is how this file is debugged, and the console is the only
  // place the answers can go.
  if (bridge) bridge.postMessage(body);
  else console.log("[engine]", body);
}

type Stage = "opening" | "reading" | "pairing" | "done";

function progress(stage: Stage, fraction: number, detail: Record<string, unknown> = {}): void {
  post({ event: "progress", stage, fraction, ...detail });
}

// ---------------------------------------------------------------------------
// Reference assets -- identical to main.ts, and deliberately so
// ---------------------------------------------------------------------------

async function loadReferenceImage(url: string): Promise<GrayImage> {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`could not load ${url} (${res.status})`);
  const bitmap = await createImageBitmap(await res.blob());

  const canvas = document.createElement("canvas");
  canvas.width = bitmap.width;
  canvas.height = bitmap.height;
  const ctx = canvas.getContext("2d", { willReadFrequently: true })!;
  ctx.drawImage(bitmap, 0, 0);
  bitmap.close();

  const { data } = ctx.getImageData(0, 0, canvas.width, canvas.height);
  return toGray(data, canvas.width, canvas.height);
}

/**
 * The handwriting model, or null if it will not load.
 *
 * Null is a supported state and not an error path: the tally counter needs
 * nothing fetched, and a scan is still worth reviewing without the digits. A
 * 3.4MB read failing on a phone at a beach is a thing that will happen, and
 * when it does the app should quietly do less rather than refuse to open.
 */
async function loadDigitModel(): Promise<DigitModel | null> {
  try {
    const res = await fetch("reference/digit-model.json");
    if (!res.ok) return null;
    return decodeModel(await res.json());
  } catch {
    return null;
  }
}

let referencesPromise: Promise<{
  images: { front: GrayImage; back: GrayImage };
  maps: { front: CellMap; back: CellMap };
  digits: DigitModel | null;
}> | null = null;

function loadReferences() {
  referencesPromise ??= (async () => {
    const [front, back, frontMap, backMap, digits] = await Promise.all([
      loadReferenceImage("reference/blank-front.png"),
      loadReferenceImage("reference/blank-back.png"),
      loadCellMap("front", async (u) => (await fetch(u)).json()),
      loadCellMap("back", async (u) => (await fetch(u)).json()),
      loadDigitModel(),
    ]);
    return { images: { front, back }, maps: { front: frontMap, back: backMap }, digits };
  })();
  return referencesPromise;
}

// ---------------------------------------------------------------------------
// What the engine holds between calls
// ---------------------------------------------------------------------------

const state = {
  fileName: "",
  fileSize: 0,
  cards: [] as ExtractedCard[],
  problems: [] as PairingProblem[],
};

/** Card 1 is column C, and the mapping is never inferred from anything else. */
function columnLetter(cardNumber: number): string {
  let n = cardNumber + 2;
  let name = "";
  while (n > 0) {
    const rem = (n - 1) % 26;
    name = String.fromCharCode(65 + rem) + name;
    n = Math.floor((n - 1) / 26);
  }
  return name;
}

function findCell(cardNumber: number, row: number): ExtractedCell | undefined {
  return state.cards.find((c) => c.cardNumber === cardNumber)?.cells.find((c) => c.row === row);
}

/**
 * Seed the date and beach from the filename, where it follows the chapter's
 * convention. Always shown for confirmation, never used silently.
 */
function seedFromFilename(name: string): { date: string; shoreline: string } | null {
  const m = /^(\d{1,2})\.(\d{1,2})\.(\d{2,4})[_-](.+?)(?:[_-]CH\d+)?\.pdf$/i.exec(name);
  if (!m) return null;
  const [, mm, dd, yy, place] = m;
  const year = yy!.length === 2 ? `20${yy}` : yy!;
  return {
    date: `${year}-${mm!.padStart(2, "0")}-${dd!.padStart(2, "0")}`,
    shoreline: place!.replace(/[-_]+/g, " ").trim(),
  };
}

// ---------------------------------------------------------------------------
// Methods
// ---------------------------------------------------------------------------

interface ProcessParams {
  /** A `cleanup://app/__inbox/…` URL Swift has just made readable. */
  url: string;
  fileName: string;
  fileSize: number;
}

/**
 * Read a scan, all the way to the cells worth showing someone.
 *
 * One page at a time, through to its crops, and then dropped -- the shape is
 * `processFile` in `main.ts` and the reason is the same: the obvious version
 * holds two full-resolution copies of the whole scan and measured 840MB of
 * browser heap on a 116-page file. That mattered on a laptop. On a phone it
 * decides whether the app survives the scan at all.
 */
async function process(params: ProcessParams) {
  progress("opening", 0);
  const refs = await loadReferences();
  const targets = referenceTargets(refs.images, refs.maps);

  const res = await fetch(params.url);
  if (!res.ok) throw new Error(`could not read the scan (${res.status})`);
  const file = new File([await res.blob()], params.fileName, { type: "application/pdf" });

  state.fileName = params.fileName;
  state.fileSize = params.fileSize;

  const pages: PageCells[] = [];
  const pageCount = await rasterizePdf(file, ({ pageNumber, image, total }) => {
    const registered = registerAgainstBestSide(image, targets, pageNumber);
    pages.push({
      pageNumber,
      side: registered.side,
      trusted: registered.trusted,
      bannerOverlap: registered.bannerOverlap,
      cells: registered.trusted
        ? cellsForSide(registered.image, pageNumber, refs.maps[registered.side], registered.side, refs.digits)
        : [],
    });
    progress("reading", (pageNumber / total) * 0.98, { pageNumber, total });
  });

  if (pageCount === 0) throw new Error("That PDF has no pages.");

  progress("pairing", 0.99);
  const { cards, problems } = pairIntoCards(pages);
  state.cards = cards.map(assembleCard);
  state.problems = problems;
  progress("done", 1);

  return {
    fileName: state.fileName,
    fileSize: state.fileSize,
    pageCount,
    seeded: seedFromFilename(params.fileName),
    /**
     * Every page, whether it was trusted or not, with the figure the refusal
     * was made on. Screen 5 shows the reviewer the real overlap rather than a
     * generic failure, which is the difference between "it did not work" and
     * "retake it in flatter light".
     */
    pages: pages.map((p) => ({
      pageNumber: p.pageNumber,
      side: p.side,
      trusted: p.trusted,
      bannerOverlap: p.bannerOverlap,
    })),
    minBannerOverlap: MIN_BANNER_OVERLAP,
    problems: state.problems,
    cards: state.cards.map((card) => ({
      cardNumber: card.cardNumber,
      column: columnLetter(card.cardNumber),
      missingSides: card.missingSides,
      cells: card.cells.map((cell) => {
        const prefill = prefillFor(cell);
        return {
          row: cell.row,
          itemName: cell.itemName,
          section: cell.section,
          side: cell.side,
          hasValue: cell.hasValue,
          tallyOnly: cell.tallyOnly,
          pageNumber: cell.pageNumber,
          // Null only where both readers declined outright; the gate is 0 and
          // fills everything either of them offers. See lib/prefill.ts.
          prefill: prefill && {
            value: prefill.value,
            confidence: prefill.confidence,
            source: prefill.source,
            // Whether this cell is taken as read and kept off the review list.
            // Decided here rather than by comparing confidences on the far
            // side, because which cells a person is shown must not depend on
            // which front end they opened -- the desktop tool and the phone
            // read the threshold from the same constant.
            autoAccepted: isAutoAccepted(prefill),
          },
        };
      }),
    })),
  };
}

type CropKind = "total" | "marks" | "context";

/**
 * One cell's picture, as a PNG.
 *
 * The enlargements are the desktop tool's, unchanged: 2.4x on the TOTAL box,
 * which is what `renderCell` shows and what the caption on screen 6 claims;
 * 2.0x and 30% wider for a tally-only row, whose number is not there to read;
 * 1.0x for the marks themselves, because counting them is easier at the size
 * they were drawn.
 */
function crop(params: { cardNumber: number; row: number; kind: CropKind }) {
  const cell = findCell(params.cardNumber, params.row);
  if (!cell) throw new Error(`no cell for card ${params.cardNumber} row ${params.row}`);

  let canvas: HTMLCanvasElement;
  switch (params.kind) {
    case "marks":
      canvas = cropToCanvas(
        cell.image,
        cell.tallyRect.x, cell.tallyRect.y, cell.tallyRect.width, cell.tallyRect.height, 1.0,
      );
      break;
    case "context":
      // 2x. The row is shown as a 34pt strip beneath the box, scrolled
      // sideways rather than squeezed to the screen's width, so it needs
      // enough pixels to stand being read at that height and no more --
      // every one of them crosses the bridge as base64.
      canvas = cropToCanvas(
        cell.image,
        cell.contextRect.x, cell.contextRect.y, cell.contextRect.width, cell.contextRect.height, 2.0,
      );
      break;
    default:
      {
        // `viewRect`, not `cell.rect`: the box plus the room a hand actually
        // uses. Cropping to the printed box cut the foot off most numbers --
        // see the note beside VIEW_MARGIN in lib/extract.ts.
        const view = viewRect(cell, cell.tallyOnly);
        canvas = cropToCanvas(
          cell.image, view.x, view.y, view.width, view.height, cell.tallyOnly ? 2.0 : 2.4,
        );
      }
  }

  return {
    width: canvas.width,
    height: canvas.height,
    // Base64 rather than a data URL: Swift wants bytes, and the `data:image/png;base64,`
    // prefix would only be stripped again on the other side.
    png: canvas.toDataURL("image/png").replace(/^data:image\/png;base64,/, ""),
  };
}

interface ExportParams {
  event: {
    date: string;
    shoreline: string;
    volunteers: number | null;
    pounds: number | null;
    durationHours: number | null;
    dataEntryVolunteer: string | null;
    club: string | null;
  };
  /** cardNumber -> [row, value][]. The typed values, and only those. */
  values: [number, [number, number][]][];
  /**
   * The values still standing as the tool put them, as [card, row] pairs.
   *
   * Swift owns this rather than the engine, because it is a record of what a
   * person has touched and the engine never sees a keystroke. A box that came
   * pre-filled and was then edited is not in here, and is exported as human.
   */
  prefilled: [number, number][];
  confidences: [number, number, number][];
}

async function exportWorkbook(params: ExportParams) {
  const res = await fetch("template/data-entry-template.xlsx");
  if (!res.ok) throw new Error("could not load the chapter's Excel template");
  const template = new Uint8Array(await res.arrayBuffer());

  const untouched = new Set(params.prefilled.map(([card, row]) => `${card}:${row}`));
  const confidence = new Map(params.confidences.map(([card, row, c]) => [`${card}:${row}`, c]));

  const cards: ExportCard[] = [];
  for (const [cardNumber, rows] of params.values) {
    if (rows.length === 0) continue;
    const source = state.cards.find((c) => c.cardNumber === cardNumber);
    cards.push({
      cardNumber,
      pageNumbers: [...new Set(source?.cells.map((c) => c.pageNumber) ?? [])].sort((a, b) => a - b),
      cardType: source?.cells.some((c) => c.tallyOnly) ? "tally" : "total",
      values: rows.map(([row, value]) => {
        // A machine reading nobody touched carries the confidence it was
        // pre-filled at and is recorded as recognized; a value a person typed
        // or corrected is certain and is recorded as human. The chapter's audit
        // column is the whole reason the two are kept apart.
        const key = `${cardNumber}:${row}`;
        const machine = untouched.has(key);
        return { row, value, confidence: machine ? (confidence.get(key) ?? 1) : 1, corrected: !machine };
      }),
    });
  }

  const bytes = fillTemplate(template, {
    sourcePdfName: state.fileName,
    event: params.event,
    cards,
  });

  let binary = "";
  for (let i = 0; i < bytes.length; i += 0x8000) {
    binary += String.fromCharCode(...bytes.subarray(i, i + 0x8000));
  }

  return {
    filename: suggestFilename(params.event.date, params.event.shoreline),
    xlsx: btoa(binary),
  };
}

/** Let go of a scan's crops. Called when the app leaves an event. */
function reset() {
  state.cards = [];
  state.problems = [];
  state.fileName = "";
  state.fileSize = 0;
  return { ok: true };
}

const METHODS: Record<string, (params: any) => unknown | Promise<unknown>> = {
  open: async () => {
    await loadReferences();
    return { ready: true };
  },
  process,
  crop,
  export: exportWorkbook,
  reset,
};

// ---------------------------------------------------------------------------

window.tally = {
  dispatch(json: string) {
    let id = 0;
    void (async () => {
      try {
        const req = JSON.parse(json) as { id: number; method: string; params?: unknown };
        id = req.id;
        const method = METHODS[req.method];
        if (!method) throw new Error(`no such method: ${req.method}`);
        post({ id, ok: true, result: await method(req.params ?? {}) });
      } catch (err) {
        post({ id, ok: false, error: err instanceof Error ? err.message : String(err) });
      }
    })();
  },
};

// Say so as soon as the modules are parsed. Swift waits for this before it
// sends anything, so a message that arrives first would be dropped.
post({ event: "loaded" });

File: ios/SurfriderDataCards/Tally/Engine.swift
//
//  The reading pipeline, driven from Swift.
//
//  There is no Swift implementation of the reading in this app and there should
//  never be one. Registration, tally counting and digit recognition are
//  measured in HANDOFF.md against the TypeScript in src/, on 1,606 pages, and a
//  port would be a second implementation to keep in step with a set of figures
//  that took months to establish. So the same modules run here, in a WKWebView
//  with no interface attached: src/engine.ts is the other half of this file.
//
//  A fix to the reading is therefore a change to src/, followed by
//  ios/sync-web.sh. It is never a change to anything in this folder.
//
//  WHY THE WEB VIEW IS IN THE HIERARCHY. It is one point square at zero
//  opacity, which looks like a trick and is not: WebKit throttles timers in a
//  web view that is not part of a window, and `rasterizePdf` yields between
//  pages so that a progress bar can paint. Throttled, that yield turns a
//  116-page scan into several extra minutes of nothing. Attached and invisible,
//  it runs at full speed. See `EngineHost`, which RootView places.
//

import Foundation
import SwiftUI
import WebKit

// MARK: - What comes back

struct ScanCell: Decodable, Identifiable, Hashable {
    struct Prefill: Decodable, Hashable {
        let value: Int
        let confidence: Double
        /// "tally" | "digits" | "agreed" | "split" -- which reader spoke.
        let source: String

        /// Whether the engine took this reading as the answer, which means this
        /// cell is not on the review list and nobody will be shown it.
        ///
        /// Optional, and absent means false: the app ships a copy of the web
        /// bundle and a build against a stale one should put every cell in
        /// front of a person rather than none. `AUTO_ACCEPT` in
        /// src/lib/prefill.ts is where the threshold lives and what it costs.
        let autoAccepted: Bool?

        var takenAsRead: Bool { autoAccepted == true }

        /// What to call this box, in the reviewer's words.
        ///
        /// The reader is named because they are not worth the same and the
        /// reviewer is entitled to know which claim they are being asked to
        /// check. "read" is the weakest of the three and is named differently
        /// on purpose. Mirrors `prefillTag` in src/lib/prefill.ts.
        var tag: String {
            switch source {
            case "digits": return "read: check it"
            case "agreed": return "counted twice: check it"
            // Nothing was read in this cell. The number in the box is a
            // starting point and not a claim, and saying so is the only thing
            // that separates it from a reading on screen.
            case "placeholder": return "nothing read: type it"
            default: return "counted: check it"
            }
        }
    }

    let row: Int
    let itemName: String
    let section: String
    let side: String
    let hasValue: Bool
    let tallyOnly: Bool
    let pageNumber: Int
    let prefill: Prefill?

    var id: Int { row }
}

struct ScanCard: Decodable, Identifiable, Hashable {
    let cardNumber: Int
    /// Card 1 is column C. Never inferred from where the ink is.
    let column: String
    let missingSides: [String]
    let cells: [ScanCell]

    var id: Int { cardNumber }
}

struct ScanPage: Decodable, Hashable {
    let pageNumber: Int
    let side: String
    let trusted: Bool
    /// How well the printed section banners landed. The figure a refusal is
    /// made on, and the one screen 5 shows rather than a generic failure.
    let bannerOverlap: Double
}

struct ScanProblem: Decodable, Hashable {
    let kind: String
    let message: String
    let pages: [Int]
}

struct ScanSeed: Decodable, Hashable {
    let date: String
    let shoreline: String
}

struct ScanResult: Decodable {
    let fileName: String
    let fileSize: Int
    let pageCount: Int
    let seeded: ScanSeed?
    let pages: [ScanPage]
    let minBannerOverlap: Double
    let problems: [ScanProblem]
    let cards: [ScanCard]
}

struct CropResult: Decodable {
    let width: Int
    let height: Int
    /// Base64 PNG. Not a data URL -- the prefix would only be stripped here.
    let png: String

    var image: UIImage? {
        Data(base64Encoded: png).flatMap(UIImage.init(data:))
    }
}

struct ExportResult: Decodable {
    let filename: String
    let xlsx: String

    var bytes: Data? { Data(base64Encoded: xlsx) }
}

struct EngineProgress {
    /// "opening" | "reading" | "pairing" | "done"
    var stage: String = "opening"
    var fraction: Double = 0
    var pageNumber: Int?
    var total: Int?
}

/// A method that answers `{ ok: true }` and nothing else.
struct EngineAck: Decodable {}

enum EngineError: LocalizedError {
    case notReady
    case failed(String)

    var errorDescription: String? {
        switch self {
        case .notReady: return "The reader has not finished starting up."
        case .failed(let message): return message
        }
    }
}

// MARK: -

@MainActor
final class Engine: NSObject, ObservableObject {

    /// True once engine.html has parsed and the reference card is in memory.
    @Published private(set) var ready = false
    @Published private(set) var progress = EngineProgress()

    private var webView: WKWebView!
    private var nextID = 1
    private var pending: [Int: CheckedContinuation<Data, Error>] = [:]
    private var loadedContinuations: [CheckedContinuation<Void, Never>] = []
    private var loaded = false

    override init() {
        super.init()

        let config = WKWebViewConfiguration()
        config.setURLSchemeHandler(WebAssetSchemeHandler(), forURLScheme: WebAssetSchemeHandler.scheme)
        config.userContentController.add(self, name: "tally")

        webView = WKWebView(frame: CGRect(x: 0, y: 0, width: 1, height: 1), configuration: config)
        webView.navigationDelegate = self
        webView.isOpaque = false
        webView.isUserInteractionEnabled = false
        // Safari's Web Inspector is the only way to see inside the pipeline
        // once it is in the app. iOS 16.4 and up; the deployment target is
        // 16.0, so it is asked for rather than assumed.
        if #available(iOS 16.4, *) { webView.isInspectable = true }

        var components = URLComponents()
        components.scheme = WebAssetSchemeHandler.scheme
        components.host = WebAssetSchemeHandler.host
        components.path = "/engine.html"
        webView.load(URLRequest(url: components.url!))
    }

    /// The view RootView plants so the web view is in a window. Invisible.
    var host: some View { EngineHost(webView: webView) }

    // MARK: - Starting up

    /// Load the reference card, the cell maps and the digit model.
    ///
    /// Worth doing before the volunteer has anything to wait for: it is a few
    /// megabytes off local storage, and doing it here means the reading starts
    /// the moment a scan arrives rather than a second later.
    func open() async {
        await waitForLoad()
        let _: EngineAck? = try? await call("open", params: [String: String]())
        ready = true
    }

    private func waitForLoad() async {
        if loaded { return }
        await withCheckedContinuation { continuation in
            loadedContinuations.append(continuation)
        }
    }

    // MARK: - Work

    /// Read a scan into cards and cells.
    ///
    /// The PDF is staged in the scheme handler's inbox and fetched by the page,
    /// rather than passed as a string: a scan is tens of megabytes, and base64
    /// through `evaluateJavaScript` would be that again with a JavaScript
    /// parser in front of it.
    func process(pdf url: URL, onProgress: @escaping (EngineProgress) -> Void) async throws -> ScanResult {
        let size = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
        let token = WebAssetSchemeHandler.stage(url)
        defer { WebAssetSchemeHandler.unstage(token) }

        progressObserver = onProgress
        defer { progressObserver = nil }

        return try await call("process", params: [
            "url": "\(WebAssetSchemeHandler.scheme)://\(WebAssetSchemeHandler.host)/__inbox/\(token)",
            "fileName": url.lastPathComponent,
            "fileSize": size,
        ] as [String: Any])
    }

    /// One cell's picture. `kind` is "total", "marks" or "context".
    func crop(card: Int, row: Int, kind: String = "total") async throws -> CropResult {
        try await call("crop", params: ["cardNumber": card, "row": row, "kind": kind] as [String: Any])
    }

    func export(_ input: [String: Any]) async throws -> ExportResult {
        try await call("export", params: input)
    }

    /// Let go of a scan's crops.
    func reset() async {
        let _: EngineAck? = try? await call("reset", params: [String: String]())
    }

    // MARK: - The bridge

    private var progressObserver: ((EngineProgress) -> Void)?

    private func call<T: Decodable>(_ method: String, params: Any) async throws -> T {
        await waitForLoad()

        let id = nextID
        nextID += 1

        let body: [String: Any] = ["id": id, "method": method, "params": params]
        guard
            let json = try? JSONSerialization.data(withJSONObject: body),
            let text = String(data: json, encoding: .utf8),
            let literal = try? JSONSerialization.data(withJSONObject: [text]),
            let quoted = String(data: literal, encoding: .utf8)
        else { throw EngineError.failed("could not encode the request") }

        // `quoted` is a one-element JSON array, so dropping the brackets leaves
        // a correctly escaped JavaScript string literal. Hand-escaping this is
        // how injection bugs are written.
        let argument = String(quoted.dropFirst().dropLast())

        let data: Data = try await withCheckedThrowingContinuation { continuation in
            pending[id] = continuation
            webView.evaluateJavaScript("window.tally.dispatch(\(argument))") { [weak self] _, error in
                guard let error else { return }
                Task { @MainActor in
                    self?.pending.removeValue(forKey: id)?.resume(throwing: error)
                }
            }
        }

        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw EngineError.failed("the reader sent back something unexpected: \(error)")
        }
    }
}

// MARK: -

extension Engine: WKScriptMessageHandler {
    func userContentController(
        _ controller: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? String, let data = body.data(using: .utf8) else { return }
        receive(data)
    }

    private func receive(_ data: Data) {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }

        if let event = object["event"] as? String {
            switch event {
            case "loaded":
                loaded = true
                for continuation in loadedContinuations { continuation.resume() }
                loadedContinuations.removeAll()
            case "progress":
                let update = EngineProgress(
                    stage: object["stage"] as? String ?? "reading",
                    fraction: object["fraction"] as? Double ?? 0,
                    pageNumber: object["pageNumber"] as? Int,
                    total: object["total"] as? Int
                )
                progress = update
                progressObserver?(update)
            default:
                break
            }
            return
        }

        guard let id = object["id"] as? Int, let continuation = pending.removeValue(forKey: id) else { return }

        if object["ok"] as? Bool == true {
            let result = object["result"] ?? [String: Any]()
            if let encoded = try? JSONSerialization.data(withJSONObject: result) {
                continuation.resume(returning: encoded)
            } else {
                continuation.resume(throwing: EngineError.failed("could not read the reply"))
            }
        } else {
            continuation.resume(throwing: EngineError.failed(object["error"] as? String ?? "the reader failed"))
        }
    }
}

extension Engine: WKNavigationDelegate {
    func webView(
        _ webView: WKWebView,
        decidePolicyFor navigationAction: WKNavigationAction,
        decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
    ) {
        // Our own bundle and nothing else. "The scan never leaves the device"
        // is the tool's first promise, and this is the line where it would
        // break -- a stray link, an injected script, a typo in a future
        // version. Refused rather than quietly loaded.
        let scheme = navigationAction.request.url?.scheme
        decisionHandler(scheme == WebAssetSchemeHandler.scheme ? .allow : .cancel)
    }
}

// MARK: - The crops, kept and fetched ahead

/// The pictures screen 6 reads from, held in memory and fetched before they are
/// asked for.
///
/// A crop is cheap and a long way away: JavaScript cuts the row out of the page
/// it is still holding, encodes a PNG and posts it back as base64, and the app
/// then decodes it twice. None of that is slow on its own. Done on arrival it
/// was all in front of the tap, three times over -- the box, the row strip and
/// the marks were awaited one after another -- so every "Next" on a screen
/// somebody works for an hour spent a white card and a spinner before it showed
/// anything. That, and not the reading, is what made the app feel slow.
///
/// So the work moves behind the tap. The cell on screen is fetched, and the
/// next few are fetched after it while the reviewer is still reading this one.
/// Advancing then hits memory and paints in the same frame.
@MainActor
final class CropCache {

    struct Key: Hashable {
        let card: Int
        let row: Int
        /// "total" | "context" | "marks", as `crop` in src/engine.ts names them.
        let kind: String
    }

    private let engine: Engine

    private var images: [Key: UIImage] = [:]
    /// Insertion order, oldest first. A decoded crop is a bitmap and an event is
    /// hundreds of cells, so this is bounded rather than left to grow into the
    /// second copy of the scan the engine went to some trouble not to keep.
    private var order: [Key] = []
    private var inFlight: [Key: Task<UIImage?, Never>] = [:]

    /// Comfortably more than the window that is prefetched, and far short of an
    /// event.
    private let limit = 48

    /// Bumped by `clear`. A fetch already in the air when the scan goes belongs
    /// to a card that is no longer loaded, and must not land in the cache the
    /// next scan will read from.
    private var generation = 0

    init(engine: Engine) { self.engine = engine }

    /// What is in memory now, for a caller that cannot wait -- which is the
    /// checking screen's body, every frame.
    func cached(_ key: Key) -> UIImage? { images[key] }

    /// The picture, from memory if it is there and from the engine if it is not.
    ///
    /// Two callers wanting the same crop share one round trip: the screen asks
    /// for the cell it is showing at the same moment the prefetch the previous
    /// cell started is still in the air, and that is the common case rather than
    /// a race worth ignoring.
    func image(_ key: Key) async -> UIImage? {
        if let hit = images[key] { return hit }
        if let running = inFlight[key] { return await running.value }

        let era = generation
        let task = Task<UIImage?, Never> { [engine] in
            guard let result = try? await engine.crop(card: key.card, row: key.row, kind: key.kind) else { return nil }
            let png = result.png
            // Base64, the PNG decode, and the bitmap the renderer actually
            // draws, all off the main thread. Left alone, the last of those
            // three happens on the first frame the picture is on screen, which
            // is the one frame that must not drop.
            return await Task.detached(priority: .userInitiated) {
                guard let data = Data(base64Encoded: png), let image = UIImage(data: data) else { return nil }
                return image.preparingForDisplay() ?? image
            }.value
        }

        inFlight[key] = task
        let image = await task.value
        inFlight.removeValue(forKey: key)
        if let image, era == generation { store(key, image) }
        return image
    }

    /// The picture, or nothing where this cell has none to show. Saves the
    /// caller a branch around three parallel fetches.
    func image(_ key: Key, when needed: Bool) async -> UIImage? {
        needed ? await image(key) : nil
    }

    /// Fetch ahead and do not wait. Anything already held or already asked for
    /// falls straight through `image`.
    func prefetch(_ keys: [Key]) {
        for key in keys where images[key] == nil && inFlight[key] == nil {
            Task { _ = await image(key) }
        }
    }

    /// The crops belong to one scan and nothing in them outlives it. Same
    /// promise as the one on every screen: the pictures stay in memory and go
    /// when the event does.
    func clear() {
        generation += 1
        for task in inFlight.values { task.cancel() }
        inFlight = [:]
        images = [:]
        order = []
    }

    private func store(_ key: Key, _ image: UIImage) {
        if images[key] == nil { order.append(key) }
        images[key] = image
        while order.count > limit {
            images.removeValue(forKey: order.removeFirst())
        }
    }
}

// MARK: -

/// Plants the engine's web view in the window so WebKit does not throttle it.
private struct EngineHost: UIViewRepresentable {
    let webView: WKWebView

    func makeUIView(context: Context) -> WKWebView { webView }
    func updateUIView(_ webView: WKWebView, context: Context) {}
}
